pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java

```diff
@@ -0,0 +1,207 @@
+    return _segmentMetadata;
+  }
+
+  @Override
+  public DataSource getDataSource(String column) {
+    ColumnMetadata columnMetadata = _segmentMetadata.getColumnMetadataFor(column);
+    Preconditions.checkNotNull(columnMetadata,
+        "ColumnMetadata for " + column + " should not be null. " + "Potentially invalid column name specified.");
+    return new ImmutableDataSource(columnMetadata, _indexContainerMap.get(column));
+  }
+
+  @Override
+  public Set<String> getColumnNames() {
+    return _segmentMetadata.getSchema().getColumnNames();
+  }
+
+  @Override
+  public Set<String> getPhysicalColumnNames() {
+    return _segmentMetadata.getSchema().getPhysicalColumnNames();
+  }
+
+  @Override
+  public void prefetch(FetchContext fetchContext) {
+    _segmentDirectory.prefetch(fetchContext);
+  }
+
+  @Override
+  public void acquire(FetchContext fetchContext) {
+    _segmentDirectory.acquire(fetchContext);
+  }
+
+  @Override
+  public void release(FetchContext fetchContext) {
+    _segmentDirectory.release(fetchContext);
+  }
+
+  @Override
+  public void destroy() {
+    String segmentName = getSegmentName();
+    LOGGER.info("Trying to destroy segment : {}", segmentName);
+
+    // Remove the upsert metadata before closing the readers
+    if (_partitionUpsertMetadataManager != null) {
+      _partitionUpsertMetadataManager.removeSegment(this);
+    }
+    for (Map.Entry<String, ColumnIndexContainer> entry : _indexContainerMap.entrySet()) {
+      try {
+        entry.getValue().close();
+      } catch (IOException e) {
+        LOGGER.error("Failed to close indexes for column: {}. Continuing with error.", entry.getKey(), e);
+      }
+    }
+    try {
+      _segmentDirectory.close();
+    } catch (Exception e) {
+      LOGGER.error("Failed to close segment directory: {}. Continuing with error.", _segmentDirectory, e);
+    }
+    if (_starTreeIndexContainer != null) {
+      try {
+        _starTreeIndexContainer.close();
+      } catch (IOException e) {
+        LOGGER.error("Failed to close star-tree. Continuing with error.", e);
+      }
+    }
+    if (_pinotSegmentRecordReader != null) {
+      try {
+        _pinotSegmentRecordReader.close();
+      } catch (IOException e) {
+        LOGGER.error("Failed to close record reader. Continuing with error.", e);
+      }
+    }
+  }
+
+  @Override
+  public List<StarTreeV2> getStarTrees() {
+    return _starTreeIndexContainer != null ? _starTreeIndexContainer.getStarTrees() : null;
+  }
+
+  @Nullable
+  @Override
+  public ThreadSafeMutableRoaringBitmap getValidDocIds() {
+    return _validDocIds;
+  }
+
+  @Override
+  public GenericRow getRecord(int docId, GenericRow reuse) {
+    try {
+      if (_pinotSegmentRecordReader == null) {
+        _pinotSegmentRecordReader = new PinotSegmentRecordReader();
+        _pinotSegmentRecordReader.init(this);
+      }
+      _pinotSegmentRecordReader.getRecord(reuse, docId);
+      return reuse;
+    } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)